### PR TITLE
Use custom DH (Diffie-Hellman) 2048 bit

### PR DESCRIPTION
**Pull request by stephdl**
https://github.com/NethServer/dev/issues/6192

Create a DH group of 2048. For the first start if the DH key doesn't exist, the the default postfix 1024 is used, until it is created.

### CODE REVIEW

**Comment by gsanchietti** [comment]
This PR is not strictly related to the new TLS policy.
Can we move it to a new features if the PR get approved?

I need to speed up TLS policy QA since it's blocking the release of multiple RPMS.

**Comment by stephdl** [comment]
agree lets find a new Issue for this @gsanchietti

**Comment by stephdl** [comment]
The key is created by a systemd `simple` service, if the key doesn't exist. No renew of the key is planned

File: server/etc/e-smith/templates/etc/postfix/main.cf/30smtpauth [hunk 1]
```diff
@@ -14,6 +14,9 @@ smtpd_sasl_authenticated_header = yes
 smtpd_tls_cert_file = /etc/pki/dovecot/certs/dovecot.pem
 smtpd_tls_key_file = /etc/pki/dovecot/private/dovecot.pem
 
+# Use specific DH group 2048 bit, renewed each week by cron
+smtpd_tls_dh1024_param_file = /etc/pki/dovecot/private/postfix.dh.param
```
Comment (DavidePrincipi) [inline server/etc/e-smith/templates/etc/postfix/main.cf/30smtpauth:18]: Can Postfix start even if the file wasn't created yet?

Comment (stephdl) [reply server/etc/e-smith/templates/etc/postfix/main.cf/30smtpauth:18]: yes it uses the default postfix dh group, then once the new dh group created it is reloaded

File: server/usr/libexec/nethserver/postfix-create-DH2048 [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+/usr/bin/openssl dhparam -out /etc/pki/dovecot/private/postfix.dh.param.tmp 2048 > /dev/null 2>&1
+
+if [[ $? != 0 ]];then
+    exit 1;
+fi
+
+/usr/bin/mv -f /etc/pki/dovecot/private/postfix.dh.param.tmp /etc/pki/dovecot/private/postfix.dh.param
+/usr/bin/chown root:root /etc/pki/dovecot/private/postfix.dh.param
+/usr/bin/chmod 600 /etc/pki/dovecot/private/postfix.dh.param
```
Comment (DavidePrincipi) [inline server/usr/libexec/nethserver/postfix-create-DH2048:38]: Not needed if `umask` is set correctly before creating the file

File: server/usr/libexec/nethserver/postfix-create-DH2048 [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+# nethserver-mail-server
+#
+# Create the diffie hellman 2048 bit of postfix
+#
+# The DH group could be renewed from time to time for security concerns.
+
+/usr/bin/openssl dhparam -out /etc/pki/dovecot/private/postfix.dh.param.tmp 2048 > /dev/null 2>&1
+
+if [[ $? != 0 ]];then
+    exit 1;
```
Comment (DavidePrincipi) [inline server/usr/libexec/nethserver/postfix-create-DH2048:33]: `.tmp` file is not cleaned up. Set a `trap` to ensure it is removed on EXIT (e.g. https://github.com/NethServer/nethserver-base/blob/7c1f0e7e4e111189dcfa0f4edca868a14c32e492/root/etc/e-smith/events/actions/nethserver-certificate-upload#L32)

File: server/usr/libexec/nethserver/postfix-create-DH2048 [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+#
+
+#
+# nethserver-mail-server
+#
+# Create the diffie hellman 2048 bit of postfix
+#
+# The DH group could be renewed from time to time for security concerns.
+
+/usr/bin/openssl dhparam -out /etc/pki/dovecot/private/postfix.dh.param.tmp 2048 > /dev/null 2>&1
```
Comment (DavidePrincipi) [inline server/usr/libexec/nethserver/postfix-create-DH2048:30]: - Better to set `umask` and generate a real temporary file with `mktemp` here.
- Why redirecting stderr to /dev/null? systemd already routes stderr to syslog...

File: server/usr/libexec/nethserver/postfix-create-DH2048 [hunk 5]
```diff
@@ -27,19 +27,24 @@
...
 
 if [[ $? != 0 ]];then
     exit 1;
 fi
 
-
 /usr/bin/systemctl reload postfix
+
+# Remove error cleanup procedure handler:
+trap - EXIT
```
Comment (DavidePrincipi) [inline server/usr/libexec/nethserver/postfix-create-DH2048:50]: We want to run the clean up procedure despite the EXIT point. 

```suggestion
```

File: server/usr/libexec/nethserver/postfix-create-DH2048 (hunk 4 above)
Comment (DavidePrincipi) [reply server/usr/libexec/nethserver/postfix-create-DH2048:35]: - With `mktemp` you have a unique and safe file to write into
- An hardcoded .tmp file can be overwritten by multiple, concurrent script runs (ok, it's likely to not happen here)
- What about stderr?

**Comment by stephdl** [comment]
> What about stderr?
Reading [this]( https://superuser.com/questions/1129314/how-to-run-openssl-dhparam-quietly)  and [this](https://github.com/search?q=org%3ANethServer+dhparam&type=Code) I could state that the build output creattion is written to stderr, we see a lot of `..+...` in loges. Hence why I discard them.

However if the command fails, we have the `$?` equal to `1`

File: server/usr/libexec/nethserver/postfix-create-DH2048 [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+trap "rm -rf ${tmpdir}" EXIT
+
+
+/usr/bin/openssl dhparam -out ${tmpdir}/postfix.dh.param 2048 2> /dev/null
+
+if [[ $? != 0 ]];then
+    exit 1;
+fi
+
+/usr/bin/mv -f ${tmpdir}/postfix.dh.param /etc/pki/dovecot/private/postfix.dh.param
```
Comment (DavidePrincipi) [inline server/usr/libexec/nethserver/postfix-create-DH2048:44]:
```suggestion
/usr/bin/mv -f ${tmpfile} /etc/pki/dovecot/private/postfix.dh.param
```

File: server/usr/libexec/nethserver/postfix-create-DH2048 [hunk 7]
```diff
@@ -29,22 +29,23 @@
 
 umask	077
 
+tmpdir=$(mktemp -d /tmp/create-DH2048.XXXXXXX)
```
[the remark this answers is not in the document]
Comment (stephdl) [reply server/usr/libexec/nethserver/postfix-create-DH2048]: fixed

File: server/usr/libexec/nethserver/postfix-create-DH2048 [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+umask	077
+
+tmpdir=$(mktemp -d /tmp/create-DH2048.XXXXXXX)
+
+# If we exit early, clean up intermediate files:
+trap "rm -rf ${tmpdir}" EXIT
+
+
+/usr/bin/openssl dhparam -out ${tmpdir}/postfix.dh.param 2048 2> /dev/null
```
[the remark this answers is not in the document]
Comment (stephdl) [reply server/usr/libexec/nethserver/postfix-create-DH2048]: fixed

File: server/usr/libexec/nethserver/postfix-create-DH2048 (hunk 6 above)
Comment (stephdl) [reply server/usr/libexec/nethserver/postfix-create-DH2048]: fixed

File: server/usr/libexec/nethserver/postfix-create-DH2048 [hunk 9]
```diff
@@ -29,22 +29,23 @@
 
 umask	077
 
+tmpdir=$(mktemp -d /tmp/create-DH2048.XXXXXXX)
+
 # If we exit early, clean up intermediate files:
-trap "rm -f /etc/pki/dovecot/private/postfix.dh.param.tmp" EXIT
+trap "rm -rf ${tmpdir}" EXIT
```
[the remark this answers is not in the document]
Comment (stephdl) [reply server/usr/libexec/nethserver/postfix-create-DH2048]: fixed

**Comment by stephdl** [comment]
> Is it worth moving them to -common?

I put this code inside -server because all TLS stuff for smtpd are inside *-server rpm, however your question is good, so I tested different installation with testssl

- nethserver-mail-smarthosts (only installed)

```
[root@ns7loc15 testssl.sh]#     ./testssl.sh   -t smtp 127.0.0.1:587
./testssl.sh: line 10592: /dev/tcp/127.0.0.1/587: Connection refused
 Oops: TCP connect problem

Unable to open a socket to 127.0.0.1:587. 
Fatal error: Can't connect to "127.0.0.1:587"
Make sure a firewall is not between you and your scanning target!
```

- nethserver-mail-common (only installed + nethserver-mail-smarthosts)

```
[root@ns7loc15 testssl.sh]#     ./testssl.sh   -t smtp 127.0.0.1:587
 Start 2020-06-08 15:46:17        -->> 127.0.0.1:587 (127.0.0.1) <<--

 rDNS (127.0.0.1):       localhost.nethservertest.org.
 STARTTLS handshake failed
 STARTTLS handshake failed
 STARTTLS handshake failed
 STARTTLS handshake failed
```
Only when I install nethserver-mail-server I can test the TLS with testssl. So I guess my first choice to put the code inside nethserver-mail-server was the good, but please if you have more doubts you can  share them @DavidePrincipi

**Comment by Neustradamus** [comment]
@DavidePrincipi, @stephdl: /!\ It is missing for ejabberd (C2Sx2/S2S/HTTPSx2)

- https://github.com/NethServer/nethserver-ejabberd/blob/master/root/etc/e-smith/templates/etc/ejabberd/ejabberd.yml/30MacrosXXXXX

define_macro:
```
  'CERTFILE': "/etc/ejabberd/ejabberd.pem"
  'TLSOPTS': XXX
  'CIPHERS': XXX
  'DHFILE': "/etc/ejabberd/dhparams.pem"
```

- https://github.com/NethServer/nethserver-ejabberd/blob/master/root/etc/e-smith/templates/etc/ejabberd/ejabberd.yml/40Listen
-> 5222
```
    protocol_options: 'TLSOPTS'
    ciphers: 'CIPHERS'
    dhfile: 'DHFILE'
```

-> 5223
```
    protocol_options: 'TLSOPTS'
    ciphers: 'CIPHERS'
    dhfile: 'DHFILE'
```
-> 5280
```
    protocol_options: 'TLSOPTS'
    ciphers: 'CIPHERS'
    dhfile: 'DHFILE'
```

-> 5269
```
    protocol_options: 'TLSOPTS'
    ciphers: 'CIPHERS'
    dhfile: 'DHFILE'
```

-> 5443
```
    protocol_options: 'TLSOPTS'
    ciphers: 'CIPHERS'
    dhfile: 'DHFILE'
```

**Pull request merged**